streams/src/test/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStoreTest.java

```diff
@@ -60,4 +67,22 @@ public void shouldRemoveKeysWithNullValues() {
 
         assertThat(store.get(0), nullValue());
     }
+
+
+    @Test
+    public void shouldReturnKeysWithGivenPrefix(){
+        store = createKeyValueStore(driver.context());
+        final String value = "value";
+        final List<KeyValue<Integer, String>> entries = new ArrayList<>();
+        entries.add(new KeyValue<>(1, value));
+        entries.add(new KeyValue<>(2, value));
+        entries.add(new KeyValue<>(11, value));
+        entries.add(new KeyValue<>(13, value));
+
+        store.putAll(entries);
+        final KeyValueIterator<Integer, String> keysWithPrefix = store.prefixScan(1, new IntegerSerializer());
```

Comment:
@cadonna I found something interesting here. As you can see this creates a state store with Integer prefix. Now when I run prefixScan, I get only 1 from the iterator even though we have 11 and 13 as the keys. This is because in prefixScan, we do this: `final Bytes to = Bytes.increment(from);` which sets `to` to 2.
Because of that, it doesn't return all the keys. 
Question is, with prefix Scan we are primarily focussing on lexicographically next key which has the same prefix. That way, 11 should be returned. But with the way it has been implemented for this store, it won't work.

Reply:
Thank you for pointing this out! 
I will have a look at this in the next days.

Reply:
@cadonna , I was thinking about this today and the question that I have is what exactly does a prefix scan mean for numeric keys. For the way we have implemented, it will work even for numeric keys if we can retrieve the keys in Lexicographical manner. 
So, if the prefix is `1` and post increment, it becomes `2`, then the range in which I need to search becomes (1,2]. If my map has 1,2,11,13 then I should be able to fetch 1,11,13 as those come before 2 lexicographically. But when I test the underlying map from my test case, it returns 1,2,11,13 and that is why the range is failing, What are your thoughts?

Reply:
I spent some more time on this today. What seems to be happening here is that in InMemoryKVStore, even though we are creating a NavigableMap as the inner map and even though Bytes(Key type for the inner map) does implement Comparable, when I insert the following keys:

```final List<KeyValue<Integer, String>> entries = new ArrayList<>();
        entries.add(new KeyValue<>(1, value));
        entries.add(new KeyValue<>(2, value));
        entries.add(new KeyValue<>(11, value));
        entries.add(new KeyValue<>(13, value));
store.putAll(entries);
```
so the expectation(atleast for me) was that i should get the keys stored lexicographically ie the output of 

```
final KeyValueIterator<Integer, String> allKeys = store.all();
        while (allKeys.hasNext()){
            System.out.println(allKeys.next().key);
        }
``` 
should be `1 11 13 2` but what gets returned is `1 2 11 13` i.e the order in which we inserted. 

Just to strengthen the point, when I insert keys in the order:
```final List<KeyValue<Integer, String>> entries = new ArrayList<>();
        entries.add(new KeyValue<>(1, value));
        entries.add(new KeyValue<>(11, value));
        entries.add(new KeyValue<>(13, value));
        entries.add(new KeyValue<>(2, value));

        store.putAll(entries);
```
and when I fetch all keys, what I get back is still `1 2 11 13` i.e the natural ordering of the integer keys. And this is precisely the reason why prefix scan is returning only 1 key in the test case that I had pointed out. I did try to pass in an explicit comparator to the TreeMap but that didn't help as well. What are your thoughts on this?

Reply:
The reason, we get only `1` when we scan for prefix `1` is that the integer serializer serializes `11` and `13` in the least significant byte instead of serializing `1` in the byte before the least significant byte and `1` and `3` in the least significant byte. With the former the **byte** lexicographical order of `1 2 11 13` would be `1 2 11 13` which corresponds to the natural order of integers. With the latter the **byte** lexicographical order of `1 2 11 13` would be `1 11 13 2` which corresponds to the string lexicographical order. So the serializer determines the order of the entries and the store always returns the entries in byte lexicographical order.

You will experience a similar when you call `range(-1, 2)` on the in-memory state store in the unit test. You will get back an empty result since `-1` is larger then `2` in byte lexicographical order   
 when the `IntegerSerializer` is used. Also not the warning that is output, especially this part `... or serdes that don't preserve ordering when lexicographically comparing the serialized bytes ...`
 
I think we should clearly state this limitation in the javadocs of the `prefixScan()` as we have done for `range()`, maybe with an example. 

Currently, to get `prefixScan()` working for all types, we would need to do a complete scan (i.e. `all()`) followed by a filter, right? 

Double checking: Is my understanding correct? @ableegoldman

Reply:
Yeah, the underlying store compares the serializer bytes lexicographically, it doesn't have any concept of "Integer" or any other type. And the really tricky thing is that it scans lexicographically, which means from left to right, whereas when we serialize things we usually do so from right to left. eg `2` in binary is `10` whereas 11 in binary is `1011` and 13 is `1101`. 
The problem here is that the serialized version of 2 is a different number of bytes than the serialized form of 11/13, so the lexicographical comparator is effectively comparing digits of a different magnitude.

Reply:
Thanks @cadonna , @ableegoldman  for the detailed explanation. I understood the behaviour now.

Reply:
> The reason, we get only `1` when we scan for prefix `1` is that the integer serializer serializes `11` and `13` in the least significant byte instead of serializing `1` in the byte before the least significant byte and `1` and `3` in the least significant byte. With the former the **byte** lexicographical order of `1 2 11 13` would be `1 2 11 13` which corresponds to the natural order of integers. With the latter the **byte** lexicographical order of `1 2 11 13` would be `1 11 13 2` which corresponds to the string lexicographical order. So the serializer determines the order of the entries and the store always returns the entries in byte lexicographical order.
> 
> You will experience a similar when you call `range(-1, 2)` on the in-memory state store in the unit test. You will get back an empty result since `-1` is larger then `2` in byte lexicographical order
> when the `IntegerSerializer` is used. Also not the warning that is output, especially this part `... or serdes that don't preserve ordering when lexicographically comparing the serialized bytes ...`
> 
> I think we should clearly state this limitation in the javadocs of the `prefixScan()` as we have done for `range()`, maybe with an example.
> 
> Currently, to get `prefixScan()` working for all types, we would need to do a complete scan (i.e. `all()`) followed by a filter, right?

That is correct. That is the only way currently. 

> 
> Double checking: Is my understanding correct? @ableegoldman

I think adding a warning similar to the range() query would be good. I will do that as part of the PR. However, in this test class, adding test cases for the integer serializer won't make sense. Probably I will create another KVStore and add tests for those. Is that ok, @cadonna ?

Reply:
That sounds good!